### PR TITLE
ADD permutation feature importance function, testing and documentatio…

### PR DESCRIPTION
**Pull request by TortySivill**
…n to permutation_feature_importance branch

---
name: Pull Request for Permutation Feature Importance
about: Template for pull requests

---

## Description

- Added Permutation Feature Importance (PFI) to existing  fatf/transparency/models/feature_influence
- Added Testing for Permutation Feature Importance to fatf/transparency/models/tests/test_feature_influence
- Added documentation for Permutation Feature Imprortance to existing:
    - doc/api
    - doc/getting_started/structure
    - doc/tutorials/model-explainability
- Added PFI example to examples/transparency/xmpl_transparency_pfi

## Reason behind implementation

New PFI functionality as agreed with Kacper

## Are there any other branches related to this work?
No

## Example code
```py

import numpy as np
import fatf.transparency.models.feature_influence as ftmfi
from sklearn.linear_model import LogisticRegression

X = np.asarray([[1, 9, 9],[1, 9, 9],[1, 9, 9],[0, 9, 9],[0, 9, 9],[0, 9, 9]])
y = np.asarray([1, 1, 1, 0, 0, 0])
clf = LogisticRegression().fit(X, y)
pfi_scores = ftmfi.permutation_feature_importance(X,clf,y,repeat_number=10)
print(np.mean(pfi_scores,axis=0))

```

## Checklist
Please ensure you have done every task in this checklist.
- [X ] Created any additional unit tests required.
- [X ] All tests pass.
- [X ] Code style is consistent with the project.
- [X ] No additional Python packages except NumPy and SciPy are required.
- [X ] The code is documented.
- [X ] Appropriate API documentation, tutorials, how-to guides and user guide entries were created.

### CODE REVIEW

File: fatf/transparency/models/feature_influence.py [hunk 1]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+    return is_input_ok
+
+
+def get_scores(dataset: np.ndarray,
+               model: object,
+               target: np.ndarray,
+               as_regressor: bool,
+               scoring_metric: Optional[str] = None):
+    """
+    Returns the difference between the trained model's predictions
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:791]: A one line summary of the functionality is needed.

File: fatf/transparency/models/feature_influence.py [hunk 2]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+    on the ``dataset`` and the ground truth ``target`` according to a
+    particular ``scoring_metric``. For the exceptions description please
+    see the documentation of the
+    :func`fatf.transparency.model.feature_influence.
+    permutation_feature_importance` function.
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:796]:
```suggestion
    permutation_feature_importance` function.

```

File: fatf/transparency/models/feature_influence.py [hunk 3]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+    Parameters
+    ----------
+    dataset : numpy.ndarray
+        A dataset based on which model predictions and associated
+        predictive error will be alculated
+    model
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:802]:
```suggestion
    model : object
```

File: fatf/transparency/models/feature_influence.py [hunk 4]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        A fitted model whose predictions will be used to calculate
+        associated predictive error based on 'scoring_metric'. (Please
+        see :class:`fatf.utils.models.models.Model` class documentation for the
+        expected model object specification.)
+    target: np.ndarray
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:807]: In the documentation, please spell out the full package name:

```suggestion
    target : numpy.ndarray
```

File: fatf/transparency/models/feature_influence.py [hunk 5]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        A vector corresponding to truth values for class labels of ``dataset``
+    as_regressor: Optional[bool] = None
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:809]: Should be spelled out in the documentation:

```suggestion
    as_regressor : boolean
```

File: fatf/transparency/models/feature_influence.py [hunk 6]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        A bool variable used to signify that the ``model``
+        is a regression model. Used to inform the ``scoring_metric``
+    scoring_metric: str
+        The type of scoring measure used to evaluate the predictive perfomance
+        of ``model`` on ``dataset`` - is ``scoring_metric`` is None, the
+        function either applies scikit learn's default
+        scorer for that model if ``model`` is a scikit-learn model,
+        else applies accuracy for a classifier and max error for a regressor.
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:817]:
```suggestion
        else applies accuracy for a classifier and max error for a regressor.

```

File: fatf/transparency/models/feature_influence.py [hunk 7]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+    Returns
+    ---------------
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:819]:
```suggestion
    -------
```

File: fatf/transparency/models/feature_influence.py [hunk 8]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+    score: float
+    Float representing the result of applying the ``scoring_metric``
+    (either default or user-inputted) performed on the outputs
+    of  the``model`` predictions on the ``dataset`` and
+    the ``target`` vector."""
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:824]:
```suggestion
        Float representing the result of applying the ``scoring_metric``
        (either default or user-inputted) performed on the outputs
        of  the``model`` predictions on the ``dataset`` and
        the ``target`` vector.
    """
```

File: fatf/transparency/models/feature_influence.py [hunk 9]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        is a regression model. Used to inform the ``scoring_metric``
+    scoring_metric: str
+        The type of scoring measure used to evaluate the predictive perfomance
+        of ``model`` on ``dataset`` - is ``scoring_metric`` is None, the
+        function either applies scikit learn's default
+        scorer for that model if ``model`` is a scikit-learn model,
+        else applies accuracy for a classifier and max error for a regressor.
+    Returns
+    ---------------
+    score: float
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:820]:
```suggestion
    score : float
```

File: fatf/transparency/models/feature_influence.py [hunk 10]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        A fitted model whose predictions will be used to calculate
+        associated predictive error based on 'scoring_metric'. (Please
+        see :class:`fatf.utils.models.models.Model` class documentation for the
+        expected model object specification.)
+    target: np.ndarray
+        A vector corresponding to truth values for class labels of ``dataset``
+    as_regressor: Optional[bool] = None
+        A bool variable used to signify that the ``model``
+        is a regression model. Used to inform the ``scoring_metric``
+    scoring_metric: str
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:812]: Optional parameter should be described in the documentation like this (the full type name should be spelled out):

```suggestion
    scoring_metric : string, optional (default=None)
```

If the default is not informative, e.g., `None`, the description below should elaborate what's happening when it's `None` and what other values are possible.

File: fatf/transparency/models/feature_influence.py [hunk 11]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        will be computed.
+    model : object
+        A fitted model whose predictions will
+        be used to calculate PFI. (Please
+        see :class:`fatf.utils.models.models.Model`
+        class documentation for the
+        expected model object specification.)
+    target: np.ndarray
+        A vector containing the true labels of the ``dataset``.
+    as_regressor: Optional[bool] = None
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:938]:
```suggestion
    as_regressor : boolean, optional (default=false)
```

File: fatf/transparency/models/feature_influence.py [hunk 12]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        is a regression model. Used to inform the ``scoring_metric``.
+    scoring_metric: Optional[str]=None
+        The type of scoring measure used to evaluate the
+        predictive perfomance of ``model`` on ``dataset``. If
+        ``scoring_metric`` is None, the function either
+        applies scikit-learn's default
+        scorer for that model if ``model`` is a scikit-learn
+        model, else applies accuracy for a classifier and max
+        error for a regressor.
+    repeat_number: Optional[int] = None (set to 10 when None)
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:949]:
```suggestion
    repeat_number : integer, optional (default=10)
```

File: fatf/transparency/models/feature_influence.py [hunk 13]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        number of times each feature column in ``dataset`` is permuted.
+
+    Warns
+    _________
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:953]:
```suggestion
    -----
```

File: fatf/transparency/models/feature_influence.py [hunk 14]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        be used to calculate PFI. (Please
+        see :class:`fatf.utils.models.models.Model`
+        class documentation for the
+        expected model object specification.)
+    target: np.ndarray
+        A vector containing the true labels of the ``dataset``.
+    as_regressor: Optional[bool] = None
+        A boolean variable used to signify that the ``model``
+        is a regression model. Used to inform the ``scoring_metric``.
+    scoring_metric: Optional[str]=None
```
Comment (So-Cool) [inline fatf/transparency/models/feature_influence.py:941]:
```suggestion
    scoring_metric : string, optional (default=None)
```

File: doc/tutorials/model-explainability.rst [hunk 15]
```diff
@@ -358,10 +359,83 @@ misinterpreting PD results, we often overlay it on top of an ICE plot::
 Such plot presents a full picture and allows us to draw conclusions about the
 usefulness of the PD curve.
 
-----
+Permutation Feature Importance
+==============================
 
-This tutorial walked through using Individual Conditional Expectation and
-Partial Dependence to explain influence of features on predictions of a model.
+Permutation Feature Importance (PFI) tells us by how much does the model's
```
[the remark this answers is not in the document]
Comment (perellonieto) [reply doc/tutorials/model-explainability.rst:365]: Agree, could be just "PFI tells us..."

File: fatf/transparency/models/feature_influence.py [hunk 16]
```diff
@@ -623,3 +660,407 @@ def partial_dependence(dataset: np.ndarray,
...
+        score = scorer(model, dataset, target)
+    return score
+
+
+def permutation_feature_importance(dataset: np.ndarray,
+                                   model: object,
+                                   target: np.ndarray,
+                                   as_regressor: Optional[bool] = None,
+                                   scoring_metric: Optional[str] = None,
+                                   repeat_number: Optional[int] = None):
```
[the remark this answers is not in the document]
Comment (perellonieto) [reply fatf/transparency/models/feature_influence.py:854]: About the mutable parameters, shouldn't it be the function who takes care of not modifying a mutable object, if that is not the intended behaviour of the function?

File: fatf/transparency/models/feature_influence.py [hunk 17]
```diff
@@ -623,3 +644,406 @@ def partial_dependence(dataset: np.ndarray,
...
+    '''
+    Calculates the Permutation Feature Importance (PFI)
+    of each feature in a dataset.
+
+    Based on the provided ``dataset``, ``model``
+    and ground truth ``target`` vector, this function computes the
+    Permutation Feature Importance (PFI).
+    PFI works by
+    permuting the values of each feature and measuring
+    the change in prediction error compared to the original
```
Comment (perellonieto) [inline fatf/transparency/models/feature_influence.py:864]: If we decide to change "prediction error" to "predictive performance metric" this needs to be changed here accordingly.

File: fatf/transparency/models/feature_influence.py [hunk 18]
```diff
@@ -623,3 +644,406 @@ def partial_dependence(dataset: np.ndarray,
...
+    ---------------
+    score: float
+    Float representing the result of applying the ``scoring_metric``
+    (either default or user-inputted) performed on the outputs
+    of  the``model`` predictions on the ``dataset`` and
+    the ``target`` vector."""
+    if SKLEARN_MISSING is True:
+        if as_regressor:
+            predictions = model.predict(dataset)  # type: ignore
+            score = -np.max(np.abs(target - predictions))
```
Comment (perellonieto) [inline fatf/transparency/models/feature_influence.py:828]: It is not clear to me if "max" should be replace by "mean". Maybe you have a better understanding of why it is more appropriate to use "max" here.

File: fatf/transparency/models/feature_influence.py [hunk 19]
```diff
@@ -623,3 +644,406 @@ def partial_dependence(dataset: np.ndarray,
...
+    If scikit-learn is not installed:
+        The user is unable to input a ``scoring_metric``.
+        If ``as_regressor`` is specified
+        the ``model`` will be treated accordingly and
+        the scoring metric applied
+        will either be ``accuracy`` for classifiers or ``max_error``
+        for regressors. If the ``as_regressor`` parameter is unspecified,
+        the model will be treated as a classifier and ``accuracy``
+        will be used to generate scores.
+
```
Comment (perellonieto) [inline fatf/transparency/models/feature_influence.py:902]: The current method to decide the "scoring_metric" sounds very convoluted. I wonder if we should simplify this in some way. Eg. expect a function that expects *ground_truth* and *scores*. Or check if the model has a scoring method.
